source/exe/win32/service_base.cc

```diff
@@ -0,0 +1,181 @@
+  // message so the user can diagnose.
+  try {
+    main_common = std::make_shared<Envoy::MainCommon>(args);
+    Envoy::Server::Instance* server = main_common->server();
+    if (!server->options().signalHandlingEnabled()) {
+      // This means that the Envoy has not registered `ENVOY_SIGTERM`.
+      // we need to manually enable it as we are going to use it to
+      // handle close requests from `SCM`.
+      sigterm_ = server->dispatcher().listenForSignal(ENVOY_SIGTERM, [server]() {
+        ENVOY_LOG_MISC(warn, "caught ENVOY_SIGTERM");
+        server->shutdown();
+      });
+    }
+  } catch (const Envoy::NoServingException& e) {
+    return S_OK;
+  } catch (const Envoy::MalformedArgvException& e) {
+    return E_INVALIDARG;
+  } catch (const Envoy::EnvoyException& e) {
+    ENVOY_LOG_MISC(warn, "Envoy failed to start with {}", e.what());
+    return E_FAIL;
+  }
+
+  return main_common->run() ? S_OK : E_FAIL;
+}
+
+void ServiceBase::Stop(DWORD control) {
+  auto handler = Event::eventBridgeHandlersSingleton::get()[ENVOY_SIGTERM];
+  if (!handler) {
+    PANIC("No handler is registered to stop gracefully, aborting the program.");
+  }
+
+  char data[] = {'a'};
+  Buffer::RawSlice buffer{data, 1};
+  auto result = handler->writev(&buffer, 1);
+  RELEASE_ASSERT(result.rc_ == 1,
+                 fmt::format("failed to write 1 byte: {}", result.err_->getErrorDetails()));
+}
+
+void ServiceBase::UpdateState(DWORD state, HRESULT errorCode, bool serviceError) {
+  status_.dwCurrentState = state;
+  if (FAILED(errorCode)) {
+    if (!serviceError) {
+      status_.dwWin32ExitCode = Win32FromHResult(errorCode);
+    } else {
+      status_.dwWin32ExitCode = ERROR_SERVICE_SPECIFIC_ERROR;
+      status_.dwServiceSpecificExitCode = errorCode;
+    }
+  }
+  SetServiceStatus();
+}
+
+void ServiceBase::SetServiceStatus() {
+  RELEASE_ASSERT(service_static != nullptr, "Global pointer to service should not be null");
+  if (!::SetServiceStatus(handle_, &status_)) {
+    PANIC(
+        fmt::format("Could not start StartServiceCtrlDispatcher with error {}", ::GetLastError()));
+  }
+}
+
+void WINAPI ServiceBase::ServiceMain(DWORD argc, LPSTR* argv) {
+  RELEASE_ASSERT(service_static != nullptr, "Global pointer to service should not be null");
+  if (argc < 1 || argv == 0 || argv[0] == 0) {
+    service_static->UpdateState(SERVICE_STOPPED, E_INVALIDARG, true);
+  }
+
+  service_static->handle_ = ::RegisterServiceCtrlHandlerA("ENVOY\0", Handler);
+  if (service_static->handle_ == 0) {
+    service_static->UpdateState(SERVICE_STOPPED, ::GetLastError(), false);
+  }
+
+  // Windows Services can get their arguments in two different ways
+  // 1. With arguments coming from StartServiceA.
+  // 2. With command line arguments that have been registered when the service gets created.
+  // We merge the two cases into one vector of arguments that we provide to main common.
```

Comment:
To make this more legible, can you invert 1 and 2? I the code should and will concat StartService args to the command line args, since the start service args are more ephemeral and should override any earlier command line args fixed in the service configuration.

Reply:
yeah makes sense, fixed in https://github.com/envoyproxy/envoy/pull/14352/commits/44ce57b0a84e1c80792ff756a5cf530a930eb15d